File: .skills/openclaw-skills/skills/gandli-2025/claw-compactor/scripts/lib/fusion/quantum_lock.py
"""Quantum Lock — KV Cache alignment as a FusionStage.

Runs at order=3, just before Cortex (order=5), so that downstream stages
always receive a prefix-stable system message.

The Anthropic prompt cache keys on the first N tokens of the system prompt.
Any dynamic content (dates, UUIDs, API keys, JWTs, timestamps) that appears
near the top of a system message will bust the cache on every request.

QuantumLock solves this by:
  1. Detecting all dynamic fragments using regex patterns.
  2. Replacing each occurrence with a stable placeholder token.
  3. Appending a clearly delimited "dynamic context" block at the END of the
     message so the model still has access to the real values.

The "quantum" metaphor: dynamic values are collapsed into a deterministic
tail section so the wavefunction of the prefix stays locked (stable).

Part of claw-compactor Phase 5. License: MIT.
"""
from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from typing import Any

from lib.fusion.base import FusionContext, FusionResult, FusionStage
from lib.tokens import estimate_tokens


# ---------------------------------------------------------------------------
# Dynamic content patterns
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class DynamicPattern:
    """A compiled pattern that identifies dynamic content."""
    name: str
    regex: re.Pattern
    placeholder: str


_RAW_PATTERNS: list[tuple[str, str, str]] = [
    # ISO 8601 date/datetime
    (
        "iso_date",
        r"\b\d{4}-\d{2}-\d{2}"
        r"(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?\b",
        "<date>",
    ),
    # Plain HH:MM:SS times
    (
        "time",
        r"\b\d{2}:\d{2}:\d{2}\b",
        "<time>",
    ),
    # JWTs (eyJ...) — header.payload.signature
    (
        "jwt",
        r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b",
        "<jwt>",
    ),
    # API keys: sk-..., rk-... OR pk_live_..., pk_test_... (Stripe-style underscore separator)
    (
        "api_key",
        r"\b(?:(?:sk|rk)-[A-Za-z0-9_-]{16,}|(?:pk_live|pk_test)_[A-Za-z0-9_-]{16,})\b",
        "<api_key>",
    ),
    # UUIDs (case-insensitive)
    (
        "uuid",
        r"\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}"
        r"-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b",
        "<uuid>",
    ),
    # Unix timestamps: 10-digit (seconds since ~2001) or 13-digit (ms)
    (
        "unix_ts",
        r"\b(?:1[5-9]\d{8}|[2-9]\d{9}|\d{13})\b",
        "<timestamp>",
    ),
    # High-entropy hex strings: 32–64 hex chars (request/trace/session IDs)
    (
        "hex_id",
        r"\b[0-9a-fA-F]{32,64}\b",
        "<id>",
    ),
]

DYNAMIC_PATTERNS: list[DynamicPattern] = [
    DynamicPattern(
        name=name,
        regex=re.compile(pattern),
        placeholder=placeholder,
    )
    for name, pattern, placeholder in _RAW_PATTERNS
]

APPENDIX_START = "<!-- quantum-lock: dynamic context -->"
APPENDIX_END = "<!-- end quantum-lock -->"
APPENDIX_SEPARATOR = "---"


# ---------------------------------------------------------------------------
# Public functions (usable standalone, not only as a FusionStage)
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class DynamicFragment:
    """A single dynamic fragment extracted from content."""
    name: str
    original: str
    placeholder: str
    indices: tuple[int, ...]  # positions in the original string


def extract_dynamic(content: str) -> list[DynamicFragment]:
    """Return all dynamic fragments found in *content*, sorted by first index.

    De-duplicates by original value: the same UUID appearing multiple times
    is reported once with all its positions.
    """
    seen: dict[str, DynamicFragment] = {}

    for dp in DYNAMIC_PATTERNS:
        for match in dp.regex.finditer(content):
            val = match.group(0)
            if val in seen:
                frag = seen[val]
                seen[val] = DynamicFragment(
                    name=frag.name,
                    original=frag.original,
                    placeholder=frag.placeholder,
                    indices=(*frag.indices, match.start()),
                )
            else:
                seen[val] = DynamicFragment(
                    name=dp.name,
                    original=val,
                    placeholder=dp.placeholder,
                    indices=(match.start(),),
                )

    return sorted(seen.values(), key=lambda f: f.indices[0])


def stabilize(content: str) -> str:
    """Stabilise *content* for KV cache alignment.

    Replaces dynamic fragments with placeholders and appends a
    "dynamic context" appendix at the end so the model still has
    access to the real values.

    Returns *content* unchanged if no dynamic fragments are found.
    """
    fragments = extract_dynamic(content)
    if not fragments:
        return content

    stabilized = content
    # Process longest originals first to avoid partial substitution
    for frag in sorted(fragments, key=lambda f: len(f.original), reverse=True):
        stabilized = stabilized.replace(frag.original, frag.placeholder)

    appendix_lines = [
        "",
        APPENDIX_SEPARATOR,
        APPENDIX_START,
    ]
    for frag in fragments:
        appendix_lines.append(f"{frag.name}: {frag.original}")
    appendix_lines.append(APPENDIX_END)

    return stabilized + "\n".join(appendix_lines)


def get_prefix_hash(content: str) -> str:
    """Return a SHA-256 hex digest of the stable prefix of *content*.

    The stable prefix is the portion before the quantum-lock appendix
    delimiter.  Identical hashes across requests indicate a likely
    prompt-cache hit.
    """
    stabilized = stabilize(content)
    marker = f"\n{APPENDIX_SEPARATOR}\n{APPENDIX_START}"
    idx = stabilized.find(marker)
    prefix = stabilized[:idx] if idx != -1 else stabilized
    return hashlib.sha256(prefix.encode("utf-8")).hexdigest()


# ---------------------------------------------------------------------------
# FusionStage implementation
# ---------------------------------------------------------------------------

class QuantumLock(FusionStage):
    """KV cache alignment stage for the Fusion Pipeline.

    Runs at order=3 (before Cortex at order=5) so that every downstream
    stage receives a prefix-stable version of the content.

    Only applies to system-role content; user/assistant/tool messages are
    passed through unchanged (they are not cached by Anthropic).
    """

    name = "quantum_lock"
    order = 3  # runs before Cortex (order=5)

    def should_apply(self, ctx: FusionContext) -> bool:
        """Apply only to system messages that contain dynamic content."""
        if ctx.role != "system":
            return False
        return bool(extract_dynamic(ctx.content))

    def apply(self, ctx: FusionContext) -> FusionResult:
        original_tokens = estimate_tokens(ctx.content)
        stabilized = stabilize(ctx.content)
        compressed_tokens = estimate_tokens(stabilized)

        fragments = extract_dynamic(ctx.content)
        markers = [
            f"quantum_lock:{frag.name}={frag.placeholder}"
            for frag in fragments
        ]

        warnings: list[str] = []
        if compressed_tokens > original_tokens:
            warnings.append(
                f"quantum_lock: stabilized content is larger than original "
                f"({compressed_tokens} > {original_tokens} tokens) — "
                f"dynamic appendix overhead"
            )

        return FusionResult(
            content=stabilized,
            original_tokens=original_tokens,
            compressed_tokens=compressed_tokens,
            markers=markers,
            warnings=warnings,
        )

File: .skills/openclaw-skills/skills/gandli-2025/claw-compactor/proxy/quantum-lock.mjs
/**
 * Quantum Lock — KV Cache alignment for Claude Code Proxy.
 *
 * The Anthropic prompt cache keys on the first N tokens of system messages.
 * Dynamic content (dates, UUIDs, API keys, timestamps) anywhere near the
 * beginning of a system message destroys cache hits every request.
 *
 * QuantumLock stabilises system messages by:
 *   1. Extracting dynamic content (dates, UUIDs, JWTs, API keys, etc.)
 *   2. Moving extracted fragments to an appendix at the END of the message
 *   3. Keeping the stable prefix identical across requests → cache hits
 *
 * The "quantum" metaphor: dynamic values are collapsed into a deterministic
 * tail section so the wavefunction of the prefix stays locked (stable).
 *
 * Part of claw-compactor Phase 5. License: MIT.
 */

import { createHash } from "node:crypto";

// ---------------------------------------------------------------------------
// Dynamic content patterns
// ---------------------------------------------------------------------------

/**
 * Each pattern entry:
 *   name    — label used in the appendix comment
 *   regex   — global RegExp for matching
 *   replace — replacement string left in the prefix position
 */
const DYNAMIC_PATTERNS = [
  // ISO 8601 dates: 2026-03-17, 2025-12-01T10:30:00Z, 2024-01-01T00:00:00.000Z
  {
    name: "iso_date",
    regex:
      /\b\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?\b/g,
    replace: "<date>",
  },

  // Plain times: 10:30:00, 23:59:59
  {
    name: "time",
    regex: /\b\d{2}:\d{2}:\d{2}\b/g,
    replace: "<time>",
  },

  // JWTs: eyJ... (base64url header.payload.signature)
  {
    name: "jwt",
    regex: /\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b/g,
    replace: "<jwt>",
  },

  // API keys: sk-..., rk-... OR pk_live_..., pk_test_... (Stripe-style underscore separator)
  {
    name: "api_key",
    regex: /\b(?:(?:sk|rk)-[A-Za-z0-9_-]{16,}|(?:pk_live|pk_test)_[A-Za-z0-9_-]{16,})\b/g,
    replace: "<api_key>",
  },

  // UUIDs: 8-4-4-4-12 hex
  {
    name: "uuid",
    regex:
      /\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/gi,
    replace: "<uuid>",
  },

  // Unix timestamps: 10-digit (seconds) or 13-digit (ms) standalone numbers
  {
    name: "unix_ts",
    regex: /\b(?:1[5-9]\d{8}|[2-9]\d{9}|\d{13})\b/g,
    replace: "<timestamp>",
  },

  // High-entropy hex strings: 32–64 hex chars (request IDs, trace IDs, session tokens)
  {
    name: "hex_id",
    regex: /\b[0-9a-f]{32,64}\b/gi,
    replace: "<id>",
  },
];

// ---------------------------------------------------------------------------
// extractDynamic
// ---------------------------------------------------------------------------

/**
 * Scan a system message for dynamic content fragments.
 *
 * Returns an array of matches, each containing:
 *   { name, original, replacement, index }
 *
 * Matches are de-duplicated by original value (same UUID appearing twice is
 * reported once with all indices).
 *
 * @param {string} systemMessage
 * @returns {Array<{ name: string, original: string, replacement: string, indices: number[] }>}
 */
export function extractDynamic(systemMessage) {
  if (typeof systemMessage !== "string") return [];

  const seen = new Map(); // original value -> entry

  for (const { name, regex, replace } of DYNAMIC_PATTERNS) {
    // Reset lastIndex since patterns are global and reused
    regex.lastIndex = 0;
    let match;
    while ((match = regex.exec(systemMessage)) !== null) {
      const key = match[0];
      if (seen.has(key)) {
        seen.get(key).indices.push(match.index);
      } else {
        seen.set(key, {
          name,
          original: key,
          replacement: replace,
          indices: [match.index],
        });
      }
    }
    regex.lastIndex = 0;
  }

  // Sort by first occurrence index so the appendix reflects document order
  return [...seen.values()].sort((a, b) => a.indices[0] - b.indices[0]);
}

// ---------------------------------------------------------------------------
// stabilize
// ---------------------------------------------------------------------------

/**
 * Stabilise a system message for KV cache alignment.
 *
 * Steps:
 *   1. Extract all dynamic fragments.
 *   2. Replace each occurrence in the text with its placeholder token.
 *   3. Append a clearly delimited "Dynamic context" section at the end
 *      listing the real values, so the model still has access to them.
 *
 * The stable prefix (everything before the appendix) is identical across
 * requests, maximising prompt-cache hit rates.
 *
 * Returns the original string unchanged if no dynamic content is found.
 *
 * @param {string} systemMessage
 * @returns {string}
 */
export function stabilize(systemMessage) {
  if (typeof systemMessage !== "string") return systemMessage;

  const fragments = extractDynamic(systemMessage);
  if (fragments.length === 0) return systemMessage;

  // Apply all replacements.
  // Process longest originals first to avoid partial substitution of substrings.
  const sorted = [...fragments].sort(
    (a, b) => b.original.length - a.original.length
  );

  let stabilized = systemMessage;
  for (const { original, replacement } of sorted) {
    // Escape special regex chars in the original value
    const escaped = original.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    stabilized = stabilized.replace(new RegExp(escaped, "g"), replacement);
  }

  // Build the appendix with actual dynamic values
  const appendixLines = ["", "---", "<!-- quantum-lock: dynamic context -->"];
  for (const { name, original } of fragments) {
    appendixLines.push(`${name}: ${original}`);
  }
  appendixLines.push("<!-- end quantum-lock -->");

  return stabilized + appendixLines.join("\n");
}

// ---------------------------------------------------------------------------
// getPrefixHash
// ---------------------------------------------------------------------------

/**
 * Compute a SHA-256 hash of the stable prefix of a system message.
 *
 * The "stable prefix" is the portion of the message before any dynamic
 * content appendix.  When this hash is identical across requests, it is
 * a strong signal that the prompt cache will hit.
 *
 * @param {string} systemMessage  raw (un-stabilised) system message
 * @returns {string}  hex digest (64 chars)
 */
export function getPrefixHash(systemMessage) {
  const stabilized = stabilize(systemMessage);

  // The stable prefix ends just before the quantum-lock appendix delimiter
  const appendixMarker = "\n---\n<!-- quantum-lock: dynamic context -->";
  const appendixIdx = stabilized.indexOf(appendixMarker);
  const prefix =
    appendixIdx === -1 ? stabilized : stabilized.slice(0, appendixIdx);

  return createHash("sha256").update(prefix, "utf8").digest("hex");
}

// ---------------------------------------------------------------------------
// createQuantumLock
// ---------------------------------------------------------------------------

/**
 * Create a QuantumLock instance.
 *
 * Returns an object with:
 *   extractDynamic(systemMessage) — list dynamic content fragments
 *   stabilize(systemMessage)      — move dynamic content to end, keep prefix stable
 *   getPrefixHash(systemMessage)  — hash of stable prefix for cache monitoring
 *
 * Also exposes:
 *   wrapRequest(body)             — stabilise the system message in a full
 *                                   chat completions request body (immutable)
 *
 * @returns {{ extractDynamic, stabilize, getPrefixHash, wrapRequest }}
 */
export function createQuantumLock() {
  /**
   * Stabilise the system message in a chat completions request body.
   * Returns a new body object (immutable — original is not mutated).
   *
   * @param {object} body  OpenAI-compatible request body
   * @returns {object}
   */
  function wrapRequest(body) {
    if (!body || !Array.isArray(body.messages)) return body;

    const messages = body.messages.map((msg) => {
      if (msg.role !== "system") return msg;

      if (typeof msg.content === "string") {
        const stable = stabilize(msg.content);
        return stable === msg.content ? msg : { ...msg, content: stable };
      }

      if (Array.isArray(msg.content)) {
        let changed = false;
        const newContent = msg.content.map((block) => {
          if (block.type !== "text") return block;
          const stable = stabilize(block.text);
          if (stable === block.text) return block;
          changed = true;
          return { ...block, text: stable };
        });
        return changed ? { ...msg, content: newContent } : msg;
      }

      return msg;
    });

    // Check whether anything actually changed
    const changed = messages.some((m, i) => m !== body.messages[i]);
    return changed ? { ...body, messages } : body;
  }

  return { extractDynamic, stabilize, getPrefixHash, wrapRequest };
}
